### ISSUE TITLE
[ImageClassifier] Broken `-convert-inout-to-fp16` due to recent refactoring

### ISSUE THREAD

**Issue by qcolombet** [pytorch/glow#1992]
TL;DR We need to do the conversion from fp32 to fp16 of the data tensor (load image) after we compile not before.

When we landed #1910 we missed a subtle difference in the order we do things that leads to broken graphs when using `-convert-inout-to-fp16` alone for the image-classifier.

Basically, `-convert-inout-to-fp16` is responsible for changing the input and output placeholder from float32 to float16 and for inserting the proper convert nodes in the graph.
When this conversion sees a placeholder with already the right type, it does nothing as it should.
So far so good.

This refactoring however changed the type of the input placeholder before running the conversion has been done, thus we end up with the placeholder wrongly typed and the conversion thinks things are fine.

When you look at the code itself, it's hard to see at first glance what we did differently.
The old order look like:
1. Load the input data
2. Create the loader
3. Convert the input data from fp32 to fp16
4. Convert the input placeholder from fp32 to fp16
5. Compiling the network
6. Running

The new order swapped 2 <-> 3:
1. Load the input data
3. Convert the input data from fp32 to fp16
2. Create the loader
4. Convert the input placeholder from fp32 to fp16
5. Compiling the network
6. Running

This is problematic because the loader actually uses the type of the input tensor data to create the placeholder and thus it thinks the network is already fp16 ready and that's the bug.

The challenge with the refactoring is that 2, 4, 5 needs to happen only once, whereas 3 needs to happen at each run. We decided to move it right after 1, whereas we actually need to move it just before 6.

To reproduce, just build `image-classifier` with asserts enabled use resnet50 (for instance) and `-convert-inout-to-fp16` and the verifier should assert.

**Comment by qcolombet** [comment]
Hey @jfix71!

Do you want to take a shot or should I?

**Comment by jfix71** [comment]
Whoops, good catch! I'll fix it.

**Comment by qcolombet** [comment]
Well I missed it during the review too :P.
